### PR TITLE
feat: P3 Permanent Staking Redesign

### PR DESCRIPTION
**Pull request by rplusq**
## Summary
  This PR implements the P3 permanent staking redesign, introducing non-decaying staking positions inspired by Velodrome alongside the existing ve-style locks. This critical upgrade enhances staking flexibility while maintaining
   complete backward compatibility for existing positions.

  ## Key Implementation Details

  ### Core Features
  - **Permanent staking**: Non-decaying locks that maintain constant weight (adapted from Velodrome)
  - **Two-phase conversion**: Atomic transition from decaying to permanent locks prevents double-counting
  - **Hybrid reward distribution**: Supports both decaying and permanent positions in weekly rewards
  - **Discrete duration set**: 4, 8, 12, 26, 52, 78, 104 weeks for predictable weights

  ### Technical Improvements
  - **AccessControl migration**: StakingRewardDistributor now uses role-based permissions
  - **Upgradeable contracts**: LockedTokenStaker made upgradeable for future enhancements
  - **Stack depth optimization**: Single storage pointer pattern avoids "stack too deep" without --via-ir
  - **Security hardening**: Fixed modifier ordering, removed deprecated functions, added missing events

  ## Code Attribution
  This implementation builds upon established DeFi patterns:
  - Core checkpoint system inspired by Curve's veCRV and PancakeSwap's VECake
  - Permanent staking patterns adapted from Velodrome's innovations
  - WalletConnect-specific adaptations including discrete duration sets, two-phase conversion mechanism, and LockedTokenStaker integration

  ## Recent Security Improvements
  - Removed unused `feed()` function to reduce code surface area
  - Fixed `nonReentrant` modifier ordering for proper reentrancy protection
  - Added constants replacing magic numbers (MAX_CHECKPOINT_ITERATIONS, MAX_REWARD_ITERATIONS)
  - Added events for better observability (RewardInjected, TotalSupplyCheckpointed)
  - Changed pragma from `^0.8.25` to `0.8.25` for deterministic compilation

  ## Storage Safety (ERC-7201)
  ```solidity
  bytes32 constant STORAGE_LOCATION = keccak256("walletconnect.storage.StakeWeight");
  // Namespaced storage prevents proxy upgrade corruption

  Critical Invariants

  1. No early withdrawal: lock.end > block.timestamp → withdrawal blocked
  2. Supply conservation: totalSupply == Σ(decaying_weights) + Σ(permanent_weights)
  3. Reward integrity: Σ(user_claims) ≤ tokensPerWeek[week]
  4. Checkpoint monotonicity: Timestamps always increasing

  Test Coverage

  - ✅ Fork tests: Verify upgrade safety on production Optimism state
  - ✅ Integration tests: Cover all permanent lock operations and conversions
  - ✅ Invariant tests: Comprehensive fuzz testing with CI profile (10k runs)
  - ✅ Unit tests: Comprehensive edge case coverage

  Known Limitations (Accepted)

  - 255 week checkpoint gap: System would break if no checkpoint for 5 years (extremely unlikely scenario)
  - 52 week claim window: Very old rewards become unclaimable (users incentivized to claim regularly)
  - Discrete durations only: Cannot create arbitrary lock durations (intentional design choice)

  Audit Documentation

  Comprehensive documentation for auditors in /docs/:
  - docs/AUDIT_SCOPE_P3.md: Security requirements, threat models, test priorities
  - docs/CODE_EVOLUTION.md: Code provenance and attribution
  - docs/MATH_AND_DESIGN.md: Mathematical formulas and implementations
  - docs/SECURITY_CONSIDERATIONS.md: Upgrade safety mechanisms

  Breaking Changes

  None - the feed() function was not actively used in production.

  Testing Instructions

  # Run all tests with force compilation
  forge test --force

  # Run with CI profile for thorough testing
  FOUNDRY_PROFILE=ci forge test

  # Fork test with production state
  source .optimism.env
  export OPTIMISM_RPC_URL=https://optimism-rpc.publicnode.com
  forge test --mc Fork --fork-url $OPTIMISM_RPC_URL

  Security Checklist

  - Storage layout verified (no slot collisions)
  - Reentrancy protection on all state-changing functions
  - Access control on privileged functions
  - Integer overflow protection with SafeCast
  - Time manipulation resistance (week alignment)
  - Fund recovery mechanisms tested

  🤖 Generated with https://claude.ai/code

  Fixed:
  - Removed fake "automated bot" mitigation claim
  - Corrected fuzz runs to match actual CI profile (10k, not 100k+)
  - Removed "migration guide" since feed() wasn't actually used
  - Changed "extremely unlikely scenario" for the 255 week gap

### CODE REVIEW

File: evm/src/StakingRewardDistributor.sol [hunk 1]
```diff
@@ -330,14 +416,25 @@ contract StakingRewardDistributor is Initializable, Ownable2StepUpgradeable, Ree
...
             // If userWeekCursor is iterated to be at/beyond maxClaimTimestamp
             // This means we went through all weeks that user subject to claim rewards already
             if (userWeekCursor >= maxClaimTimestamp) {
                 break;
             }
-            // Move to the new epoch if need to,
-            // else calculate rewards that user should get.
+            // Calculate balance for current week BEFORE moving to new epoch
+            // This properly handles both permanent and decaying locks
+            uint256 balanceOf = this.balanceOfAt(user, userWeekCursor);
```
[the remark this answers is not in the document]
Comment (rplusq) [reply evm/src/StakingRewardDistributor.sol:424]: not worth the code duplication
